sdk/core/azure-core/src/http/challenge_based_authentication_policy.cpp

```diff
@@ -0,0 +1,126 @@
+  auto const& rawDataHeader = headers.find("www-authenticate");
+  // Only re-run when Unauthozied AND www-authenticate header in the response
+  if (result->GetStatusCode() == HttpStatusCode::Unauthorized && rawDataHeader != headers.end())
+  {
+    ChallengeParameters challenge(rawDataHeader->second);
+    Credentials::TokenRequestContext tokenRequestContext;
+
+    tokenRequestContext.TenantId = std::move(challenge.TenantId);
+    tokenRequestContext.AuthorizationUri = std::move(challenge.AuthorizationUri);
+    tokenRequestContext.Scopes = std::move(challenge.Scopes);
+
+    {
+      std::lock_guard<std::mutex> lock(m_accessTokenMutex);
+      m_accessToken = m_credential->GetToken(tokenRequestContext, context);
+      request.SetHeader("authorization", "Bearer " + m_accessToken.Token);
+    }
+    return nextPolicy.Send(request, context);
+  }
+
+  // Did not re-send on any other case
+  return result;
+}
+/*
+ * the raw value ( the value of the "www-authenticate" header is in the following format
+ * "Bearer authorization/authorization_uri=[value] resource/scope=[value]";
+ */
+ChallengeParameters::ChallengeParameters(std::string const& headerValue)
+{
+  if (headerValue.rfind(_detail::BearerName) == 0)
+  {
+    Schema = _detail::BearerName;
+
+    std::vector<std::string> const parts = GetParts(headerValue, _detail::SpaceSeparator);
+    for (size_t i = 1; i < parts.size(); i++)
+    {
+      ProcessFragment(parts[i]);
+    }
+  }
+}
+
+void ChallengeParameters::ProcessFragment(std::string const& fragment)
+{
+  // check that fragment is in the Key=value format
+  if (fragment.rfind(_detail::EqualSeparator) != std::string::npos)
+  {
+    auto subParts = GetParts(fragment, _detail::EqualSeparator);
+
+    for (size_t i = 0; i < subParts.size(); i++)
+    {
+      // some parts (values) have quotes around , thus clean them on quotes
+      auto finalParts = GetParts(subParts[i], _detail::QuoteSeparator);
+
+      if (finalParts.size() > 0)
+      {
+        // take the first piece that has been trimmed
+        subParts[i] = finalParts[0];
+      }
+    }
+
+    // authorization or authorization_uri go to authorization field
+    if (subParts[0] == _detail::AuthorizationName || subParts[0] == _detail::AuthorizationUriName)
+    {
+      AuthorizationUri = Url(subParts[1]);
+      // auth tenant is part of the authorization uri
+      TenantId = AuthorizationUri.GetPath();
+      AuthorizationUri.AppendPath("oauth2/v2.0/token");
+    }
+    // scopes are either resource or scope
+    else if (subParts[0] == _detail::ResourceName)
+    {
+      Scopes.emplace_back(subParts[1] + _detail::DefaultSuffix);
+    }
+    else if (subParts[0] == _detail::ScopeName)
+    {
+      Scopes.emplace_back(subParts[1]);
+    }
+  }
+}
+
+std::vector<std::string> ChallengeParameters::GetParts(
+    std::string const& inputString,
+    char const& separator)
+{
+  std::vector<std::string> returnValue;
+  size_t startPosition = 0;
+  size_t endPosition = 0;
```

Comment:
use iterators linked to the source string.  The iterator holds a link to the source string and the current position.
Using integers for the position makes it complicated to use these values without the source string.
For C we use pointers to read arrays (instead of just the integer position).
For C++ we use iterators to read standard containers (instead of just the integer position).

Reply:
i use the position to find the substring that i need to make a copy of. 
it's standard code , nothing i invented. 
why do i need to overcomplicate things using iterators to get a substring ?
 if the iterators were meant to be used then the find function would return an iterator not an index in the string. since the string is an array and all i need is a copy between 2 indexes to me it seems like using a tank to kill a mosquito. 
what benefit does adding the another construct on top of an array to get the same results i already obtain?